### ISSUE TITLE
App restarts when (dis)connecting a keyboard

### ISSUE THREAD

**Issue by gersilex** [vbier/habpanelviewer#48]
**Describe the bug**
When a keyboard is connected or disconnected, the app restarts cleanly, without crashing, persisting the same PID.

**To Reproduce**
- Be anywhere in the app, WebView, Settings or Status Information
- Connect or Disconnect a USB or Bluetooth keyboard
- if not already visible, change to the WebView
- App restarts as soon as WebView is visible, connection time counter gets reset to zero in connection statistics in status information view.

**Expected behavior**
App does not restart or reconnect, input can be provided using keyboard

**Screenshots**
N/A

**Smartphone/Tablet (please complete the following information):**
 - Devices: OnePlus 6 (A6003), Alcatel 1T 10 (8082)
 - OS: Android 9 Kernel 4.9.106, Android 8.1.0 Kernel 3.18.79
 - Version 0.9.22 (Google Play), 0.9.25 (app-debug.apk)

**Additional context**
Output of `adb logcat --pid=$(adb shell pidof -s de.vier_bier.habpanelviewer)` (data is non-private, don't worry)

```log
10-09 22:24:06.791 17861 17861 V HPV-MainActivity: in destroy
10-09 22:24:07.441 17861 18638 D HPV-FlashHandler: FlashControlThread finished
10-09 22:24:07.442 17861 17861 V HPV-MainActivity: terminating camera...
10-09 22:24:07.442 17861 18639 D HPV-Camera: terminating...
10-09 22:24:07.442 17861 18639 D HPV-Camera: terminating finished
10-09 22:24:07.443 17861 17861 E SensorManager: unregisterListenerImpl callingApp: de.vier_bier.habpanelviewer,callingPid:17861,callingUid:10320
10-09 22:24:07.444 17861 17861 I chatty  : uid=10320(de.vier_bier.habpanelviewer) identical 2 lines
10-09 22:24:07.444 17861 17861 E SensorManager: unregisterListenerImpl callingApp: de.vier_bier.habpanelviewer,callingPid:17861,callingUid:10320
10-09 22:24:07.444 17861 17861 D HPV-NetworkTracker: unregistering network receiver...
10-09 22:24:07.445 17861 17861 D HPV-ClientWebView: unregistering as network listener...
10-09 22:24:07.458 17861 18640 V HPV-MotionDetector: motion thread finished
10-09 22:24:07.508 17861 17861 D HPV-MainActivity: Version: 0.9.25
10-09 22:24:07.531 17861 17861 D HPV-MainActivity: SSL protocol initialized
10-09 22:24:07.531 17861 17861 D HPV-NetworkTracker: registering network receiver...
10-09 22:24:07.532 17861 17861 D HPV-NetworkTracker: network is active: [type: WIFI[], state: CONNECTED/CONNECTED, reason: (unspecified), extra: (none), failover: false, available: true, roaming: false]
10-09 22:24:07.532 17861 17861 D HPV-NetworkTracker: notifying listeners of active network...
10-09 22:24:07.535 17861 18678 D HPV-FlashHandler: FlashControlThread started
10-09 22:24:07.543 17861 17861 D HPV-ClientWebView: registering as network listener...
10-09 22:24:07.549 17861 17861 D HPV-ClientWebView: loadStartUrl: loading start page https://openhab.nest.de:8443...
10-09 22:24:07.549 17861 17861 D HPV-ServerConnection: new server URL: https://openhab.nest.de:8443
10-09 22:24:07.549 17861 17861 D HPV-ServerConnection: EventSource connection skipped: no subscriptions
10-09 22:24:07.549 17861 18679 D HPV-Camera: updating from preferences...
10-09 22:24:07.550 17861 18679 D HPV-Camera: updating from preferences finished
10-09 22:24:07.576 17861 17874 D DecorView: onWindowFocusChangedFromViewRoot hasFocus: true, DecorView@9dc298e[MainActivity]
10-09 22:24:07.565 17861 17861 W RenderThread: type=1400 audit(0.0:112021): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:07.579 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.changepixelformat"
10-09 22:24:07.598 17861 17861 D OnePlusJankManager:  Chor uploadMDM JANK_TYPE_ONCE mViewTitle = de.vier_bier.habpanelviewer/de.vier_bier.habpanelviewer.MainActivity--- jank level = 1
10-09 22:24:07.599 17861 17861 D HPV-TrackShutdownServ: Service Started
10-09 22:24:07.599 17861 17861 D HPV-NetworkTracker: network broadcast received
10-09 22:24:07.600 17861 17861 D HPV-NetworkTracker: network is active: [type: WIFI[], state: CONNECTED/CONNECTED, reason: (unspecified), extra: (none), failover: false, available: true, roaming: false]
10-09 22:24:07.624 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:07.615 17861 17861 W RenderThread: type=1400 audit(0.0:112022): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:07.635 17861 17861 W RenderThread: type=1400 audit(0.0:112023): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:07.646 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:08.031 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:08.025 17861 17861 W RenderThread: type=1400 audit(0.0:112024): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:08.036 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:08.025 17861 17861 W RenderThread: type=1400 audit(0.0:112025): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:08.046 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:08.035 17861 17861 W RenderThread: type=1400 audit(0.0:112026): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:08.095 17861 17861 W RenderThread: type=1400 audit(0.0:112027): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:08.110 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.161 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.155 17861 17861 W RenderThread: type=1400 audit(0.0:112028): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.165 17861 17861 W RenderThread: type=1400 audit(0.0:112029): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.179 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.185 17861 17861 W RenderThread: type=1400 audit(0.0:112030): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.196 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.195 17861 17861 W RenderThread: type=1400 audit(0.0:112031): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.209 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.215 17861 17861 W RenderThread: type=1400 audit(0.0:112032): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.226 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.235 17861 17861 W RenderThread: type=1400 audit(0.0:112033): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.244 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.262 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.255 17861 17861 W RenderThread: type=1400 audit(0.0:112034): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=16891 scontext=u:r:untrusted_app:s0:c64,c257,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
10-09 22:24:09.277 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
10-09 22:24:09.413 17861 17943 I chatty  : uid=10320(de.vier_bier.habpanelviewer) RenderThread identical 8 lines
10-09 22:24:09.434 17861 17943 E libc    : Access denied finding property "vendor.debug.egl.swapinterval"
```

**Comment by gersilex** [comment]
If I can provide you anything more, please let me know.

**Comment by vbier** [comment]
I have just updated the apk in the master branch. As I have no keyboard to test this with, can you please check if this works for you?

**Comment by gersilex** [comment]
I'll test it and report back today. :+1:

**Comment by gersilex** [comment]
The new `app-debug.apk` at  https://github.com/vbier/habpanelviewer/commit/32d39c811b3349b08524cc41c87aa3bfd7cc7ec0 unfortunately did not change that behaviour at all.

**Comment by vbier** [comment]
If you really checked out 32d39c8 then it is normal that nothing changed. I forgot to commit the apk and then updated the apk in the next commit. If you took the latest version I will have to check if I can reproduce this in the simulator.

**Comment by gersilex** [comment]
I downloaded the file directly from GitHub while being on the master branch.

![image](https://user-images.githubusercontent.com/2018039/67494312-4b851780-f679-11e9-9d89-4da9b36e4132.png)

To me it really looks like you changed the Manifest *and* the APK in that commit. 

The `app-debug.apk`'s latest object version is contained in that commit:
![image](https://user-images.githubusercontent.com/2018039/67494507-93a43a00-f679-11e9-8003-4c72091563cb.png)

It's also mentioned in the diff, but collapsed because those are binary changes that cannot be diffed.

![image](https://user-images.githubusercontent.com/2018039/67494249-37411a80-f679-11e9-8ec5-73e9c98eaae7.png)

So I believe that is the latest version of the apk. And I was on the master as you instructed. I think it's time for the emulator :)

Let me know if you can reproduce the behaviour. I can test on real devices for you anytime.

**Comment by vbier** [comment]
I have now borrowed a bluetooth keyboard, fixed the problem and updated the apk once again. Can you please test if it also works for you?

**Comment by gersilex** [comment]
Great. I'll report back late this evening.

**Comment by gersilex** [comment]
@vbier This looks a lot better and works flawlessly with my Bluetooth keyboard. The issue still persists with USB keyboard. ~I might have to add the information that my devices don't have a USB-A port which I use for connecting the keyboard. Instead, I use a USB-OTG-conform adapted to connect my USB-A devices to the USB-Micro-B port on the devices.~

My USB keyboard has also a touchpad included. When connecting it, I can see that the App is handling a configuration change. I am pretty sure, this is the change emitted by connecting the keyboard part, which we now handle.

I think connecting the mouse/touchpad part of the USB keyboard is firing a `navigation` configuration change. https://developer.android.com/guide/topics/manifest/activity-element.html#config

What do think, should we handle it as well? My problem is already fixed but I think a lot people might have those combined keyboard/touchpad devices as well.

**Comment by vbier** [comment]
I have updated the apk once again. Can you retest, please?

**Comment by gersilex** [comment]
`navigation` did the trick. The Activity does not restart anymore and I see additional handling of that ActivityRestart event in the logs.

So everything runs fine now and (un)plugging of a mouse or keyboard devices, including bluetooth does not restart the Activity anymore. Thank you!

Fixed in 620ddee167e64e3b5ca57d96e69164da02457e7a
